src/osd/scheduler/mClockScheduler.cc

```diff
@@ -358,19 +387,11 @@ void mClockScheduler::set_global_recovery_options()
   cct->_conf.apply_changes(nullptr);
 }
 
-int mClockScheduler::calc_scaled_cost(int cost)
+int mClockScheduler::calc_scaled_cost(int item_cost)
 {
-  // Calculate scaled cost in msecs based on item cost
-  int scaled_cost = std::floor((cost / max_osd_bandwidth) * 1000);
-
-  // Scale the cost down by an additional cost factor if specified
-  // to account for different device characteristics (hdd, ssd).
-  // This option can be used to further tune the performance further
-  // if necessary (disabled by default).
-  if (osd_mclock_cost_per_io_msec > 0) {
-    scaled_cost *= osd_mclock_cost_per_io_msec / 1000.0;
-  }
-
+  // Calculate total scaled cost in secs
+  int scaled_cost =
+    std::round(osd_mclock_cost_per_io + (osd_mclock_cost_per_byte * item_cost));
```

Comment:
this construction is much more intuitive, thanks!

I'm confused about the hdd/ssd scaling distinction - why is one in seconds, and the other milliseconds?
For SSD, it seems like 1ms is a high minimum cost compared to the actual time it takes for an i/o. And 1s seems like a high minimum cost for HDD. What happens if we use usec for both, so that when we convert to ints for mclock we can represent smaller costs?

Reply:
This relates to a general pet peeve of mine: can we plan for a future move from
float/double for time durations, to (for C++) chrono::milliseconds (and, for configuration files and non-C++, the
number of milliseconds)?

Reply:
> this construction is much more intuitive, thanks!
> 
> I'm confused about the hdd/ssd scaling distinction - why is one in seconds, and the other milliseconds?
> For SSD, it seems like 1ms is a high minimum cost compared to the actual time it takes for an i/o. And 1s seems like a high minimum cost for HDD. What happens if we use usec for both, so that when we convert to ints for mclock we can represent smaller costs?

@jdurgin @athanatos As discussed, for HDDs and SSDs the final costs per io in secs and msecs respectively was 
determined using a simple formula (i.e. item_cost(bytes)/max_osd_bandwidth) first and then using a multiplication factor 
to arrive at the final scaled cost that worked well across the item cost range. Linear regression analysis was then performed
on the data set that worked well to finally arrive at the cost_per_io and the cost_per_byte values.

The scaled costs for SSDs were arrived at by scaling them by factor of 1000 and therefore the cost for SSDs is represented
in terms of msec.

Another factor that necessitates the different scaling between HDDs and SSDs is that the dmclock_server expects the 
"cost" parameter as part of add_request() (https://github.com/ceph/dmclock/blob/9b1326cfa09735ae18b1c4c308ec424c5f1b9cce/src/dmclock_server.h#L1388) to be an 'int'. Tag calculations are
 performed using this cost value (https://github.com/ceph/dmclock/blob/9b1326cfa09735ae18b1c4c308ec424c5f1b9cce/src/dmclock_server.h#L256). As discussed the cost may be changed to accept smaller values (in terms for msec, usec) to 
have finer grained control for items having lower cost. Right now, there are a range of lower item costs for which the cost 
parameter is set to 1 due to which proper QoS can't be provided.

Thoughts?